### DISCUSSION TITLE
Handling `wc_add_notice` on Cart/Checkout Store API routes

### DISCUSSION THREAD

**Discussion by opr** [woocommerce/woocommerce#46489] [Checkout Flow]
## Goal of discussion

The goal of this discussion is to align on a solution for extensions that need to add notices during a Store API request but aren't necessarily errors, e.g. debug type notices for merchants-facing debug information in shipping/payment methods.

## The problem

It is not possible to show notices that were added during an API request in the Cart/Checkout blocks.

>When using the classic cart/checkout shortcodes, we are able to use wc_add_notice() to display different types of notices about the cart/checkout forms as needed. For example, we can display info notices containing debug information for merchants who have shipping debug enabled. We can also display success notices, for example in cases where an address has been verified. We can also display error notices when something is not right.
>
> When using the cart/checkout blocks, we can no longer display notices the same way. The most we can do is display an error notice by throwing a WC_REST_Exception during the woocommerce_store_api_cart_update_customer_from_request hook. There is no way to display debug information in an info notice or any kind of success notices, from what we can tell.

## Solutions

1. Developers could extend the Store API routes and add notices in the `extensions` key. On the client-side, read this data and create notices using `@wordpress/data`'s `core/notices` store.
  - This seems like a lot of work for adding a simple notice. It shouldn't require a lot of code
  - Each extension would have to duplicate the same code to achieve the same result.
  - Don't see a way of this working for error responses
  - This is not really accessible to decoupled checkouts if the logic for displaying the notices is handled in the extension's code.
2. Automatically add notices in the response of each Store API request and show them
  - We'd also need notices to supply a context to determine where they'll be shown. The benefit of this is that extensions have to be intentional about what notices are shown in blocks, and where. Without supplying a valid context the notice would not show.
  - We would need to process errors on both successful _and_ error responses, therefore some code duplication would occur (we handle and generate error responses separately)
  - The server-side equivalent of an `info` notice is called `notice` so we're merging two systems with slight differences. As long as this is documented I think it's OK.
  - No `warning` type exists for server-side notices.

My preference is 2 as it makes it a lot easier for developers, and considering decoupled apps, it basically makes 1 a non-starter. The dev-ex of solution 1 is very poor as well.

## Example solution

In https://github.com/woocommerce/woocommerce/pull/46488 I made an example solution that handles notices added during a failed payment attempt. To test it out, you can download this failed payment gateway plugin and observe the notices created when using the method.
[woocommerce-failed-gateway.zip](https://github.com/woocommerce/woocommerce/files/14945043/woocommerce-failed-gateway.zip)

To change the notices, go to the `process_payment` method of `gateway-class.php`.

### Uncertainties of solution

Currently, I'm unsure if I should add notices to the cart response. There are several points during the checkout flow where the client-side cart is replaced by a cart from the server:

- Adding and removing coupons
- Updating shipping method
- Changing address
- When an extension calls `extensionCartUpdate`

The cart response already contains `errors` which are added when an exception is thrown during certain parts of the Store API request.

I also think that it's a good idea to skip adding any notices without a context passed in the `data` arg. Requiring a context makes developers think about where the notice will appear, and makes the use of `wc_add_notice` in the store API context an opt-in thing, so notices that aren't explicitly intentional won't show up.

## Feedback needed

The feedback I would like at this stage is:

1. What are your use-cases for adding notices during Store API requests that _aren't_ payment attempts? Can you provide some example steps/code of how you'd add them?
2. Should context be required? If it's not passed, should it default to the top level (`wc/cart` or `wc/checkout`)?
3. Should notices be included in the cart response?

Please provide feedback by May 3rd 2024.

**Comment by pierorocca** [comment]
Terrific writeup @opr. Too bad there are no responses. What's your recommendation on where to go from here?

**Reply by opr** [reply]
Wow 😁 yeah I forgot about this one. I'd say let's do a _research spike_ on idea number 2 and see how the dev ex looks for that as well as how easy it'll be for us to implement.

**Reply by pierorocca** [reply]
Do we have any legit use cases that aren't for "troubleshooting" which I assume is debugging related. What are the risks to scalability and interoperability if notices can proliferate vs. curating a maintained list that can be triggered?

**Reply by opr** [reply]
Coming back to look at this, the use case of 

> We can also display success notices, for example in cases where an address has been verified. We can also display error notices when something is not right.

Seems like one, however there _are_ mechanisms in place for adding notices. I think this issue is mainly to make it easier for developers to make their extensions compatible with the blocks. If their extension works 100% on the server side, asking them to go through all the effort to build a frontend integration just to show a notice is going to huge point of friction.

> What are the risks to scalability and interoperability if notices can proliferate vs. curating a maintained list that can be triggered?

I think curating a list isn't going to be possible, but I don't see any implications in terms of scalability or compatibility with doing what currently happens on the shortcode (showing all notices added with `wc_add_notice`) the reach and lifecycle of these notices doesn't go further than the checkout block.

**Reply by pierorocca** [reply]
Thoughts on an internal RFC collecting input from Rubik, Heisenberg, Quark/Fractal, and design so that we have a diverse set of informed opinions, and being inclusive of design guidelines @opr before commiting capacity?

**Reply by opr** [reply]
@mikejolley before we commit anything to this, do you have any thoughts on collecting and outputting _all_ notices added (I know we add some, on some routes, I think specifically only on fail, though?)

**Reply by mikejolley** [reply]
Is the lack of discussion on this thread an indication there isn't much need for a built in solution for store notices? Debugging information seems very niche, and there should be other mechanisms to log debugging information for merchants if needed e.g. the logs.

Additionally, if debugging is the only use case, they do not need to necessary need to render on the frontend. From memory and experience I would say that the notices were only used in this manner in the past because there were a lack of alternatives. For example, I remember we used to surface "shipping" debugging via notices (returned rates, matching zone) because there was no alternative in place.

There could also be two parts to this:

1) Including notices in store api responses
2) Surfacing them in the UI

I don't think we necessarily need to take care of both, so some real world use cases would be useful before making a decision.

> do you have any thoughts on collecting and outputting all notices added

I'd worry this would pick things up out of context. Notices can be added anywhere by anything and abused. Additionally, we would have the problem of deciding how to handle a notice in a response that causes navigation. E.g. if a notice comes back via the `/checkout/` endpoint, when is it surfaced, is it blocking, and if its not immediately shown, where is it stored.
